hooks/pre-checkout

```diff
@@ -0,0 +1,296 @@
+  local RUNTIME_OS=""
+  local RUNTIME_CPU=""
+
+  case "${UNAME_INFO}" in
+    *Darwin*)
+      RUNTIME_OS=macOS
+      ;;
+    *Linux*)
+      RUNTIME_OS=Linux
+      ;;
+    *FreeBSD*)
+      RUNTIME_OS=FreeBSD
+      ;;
+    *)
+      fail_with_message "failed to determine current computer's operating system"
+      return 1
+      ;;
+  esac
+
+  case "${UNAME_INFO}" in
+    *amd64*)
+      RUNTIME_CPU=64bit
+      ;;
+    *x86_64*)
+      RUNTIME_CPU=64bit
+      ;;
+    *arm64*)
+      RUNTIME_CPU=ARM64
+      ;;
+    *)
+      fail_with_message "failed to determine current computer's cpu"
+      return 2
+      ;;
+  esac
+
+  echo "${RUNTIME_OS}-${RUNTIME_CPU}"
+
+  return 0
+}
+
+# trivy_tar_url builds the tar.gz URL for the specified version
+# of trivy and writes it to stdout.
+#
+# Any non-zero status code indicates an error.
+#
+# For example:
+# https://github.com/aquasecurity/trivy/releases/download/v0.31.3/trivy_0.31.3_FreeBSD-64bit.tar.gz
+trivy_tar_url() {
+  local VERSION="${1}"
+  [[ -z "${VERSION}" ]] && return 1
+
+  local OS_CPU="${2}"
+  [[ -z "${OS_CPU}" ]] && return 2
+
+  local BASE_URL="https://github.com/aquasecurity/trivy/releases/download"
+  local URL="${BASE_URL}/v${VERSION}/trivy_${VERSION}_${OS_CPU}.tar.gz"
+
+  echo "${URL}"
+  return 0
+}
+
+# trivy_hashes_url builds the hashes URL for the specified version
+# of trivy and writes it to stdout.
+#
+# Any non-zero status code indicates an error.
+#
+# For example:
+# https://github.com/aquasecurity/trivy/releases/download/v0.31.3/trivy_0.31.3_checksums.txt
+trivy_hashes_url() {
+  local VERSION="${1}"
+  [[ -z "${VERSION}" ]] && return 1
+
+  local BASE_URL="https://github.com/aquasecurity/trivy/releases/download"
+  local URL="${BASE_URL}/v${VERSION}/trivy_${VERSION}_checksums.txt"
+
+  echo "${URL}"
+  return 0
+}
+
+# sha2_256_hash_file hashes the specified file using the provided
+# SHA2-256 hash string, returning status code zero on success.
+#
+# Any non-zero status code indicates an error.
+sha2_256_hash_file() {
+  local FILE="${1}"
+  [[ -z "${FILE}" ]] && return 1
+  [[ ! -f "${FILE}" ]] && return 2
+
+  local HASH="${2}"
+  [[ -z "${HASH}" ]] && return 3
+
+  local RESULT=""
+  if which sha256sum 2>&1 > /dev/null; then
+    # $ RESULT="$(sha256sum /proc/self/cmdline)"
+    # $ echo "${RESULT}"
+    # 31d14183b4... /proc/self/cmdline
+    # $ RESULT="${RESULT% *}"
+    # $ echo "${RESULT}"
+    # 31d14183b4...
+    RESULT="$(sha256sum "${FILE}")"
+    RESULT="${RESULT% *}"
+  elif which openssl 2>&1 > /dev/null; then
+    # $ RESULT="$(openssl sha256 /proc/self/cmdline)"
+    # $ echo "${RESULT}"
+    # SHA256(/proc/self/cmdline)= fbc0b267807d5d...
+    # $ RESULT="${RESULT##* }"
+    # $ echo "${RESULT}"
+    # fbc0b267807d5d...
+    RESULT="$(openssl sha256 "${FILE}")"
+    RESULT="${RESULT##* }"
+  else
+    return 4
+  fi
+
+  [[ -z "${RESULT}" ]] && return 41
+
+  [[ "${HASH}" == "${RESULT}" ]] && return 0
+
+  return 66
```

Comment:
if we're gonna be managing return codes, we should document what they mean to make it easier for folks to debug. alternatively, we should have a function that takes a return code and translates that into a human-readable message.

Reply:
I ended up making the script exit immediately in places it was returning a non-zero exit status code.

Reply:
That's fair enough as well. As long as we return an understandable error message, we're good.